sys/mips/include/atomic.h

```diff
@@ -882,6 +712,232 @@ atomic_fetchadd_64(__volatile uint64_t *p, uint64_t v)
 }
 #endif
 
+#undef ATOMIC_CMPSET
+#undef ATOMIC_FCMPSET
+
+#ifdef __CHERI_PURE_CAPABILITY__
+/*
+ * CHERI-only variants to perform atomic operations on pointers.
+ */
+
+static __inline void
+atomic_set_ptr(volatile uintptr_t *p, uintptr_t v)
+{
+	uintptr_t temp;
+
+	__asm __volatile (
+		"1:\n\t"
+		"cllc		%0, %1\n\t"	/* load old value */
+		"cgetaddr	$t0, %0\n\t"	/* calculate new value */
+		"cgetaddr	$t1, %2\n\t"
+		"or		$t0, $t1, $t0\n\t"
+		"csetaddr	%0, %0, $t0\n\t"
+		"cscc		$t0, %0, %1\n\t" /* attempt to store */
+		"beqz		$t0, 1b\n\t"	/* spin if failed */
+		"nop\n\t"			/* delay slot */
+		: "=&r" (temp), "+C" (p)
+		: "r" (v)
+		: "t0", "t1", "memory");
+}
+
+static __inline void
+atomic_clear_ptr(volatile uintptr_t *p, uintptr_t v)
+{
+	uintptr_t temp;
+	v = ~v;
+
+	__asm __volatile (
+		"1:\n\t"
+		"cllc		%0, %1\n\t"	/* load old value */
+		"cgetaddr	$t0, %0\n\t"	/* calculate new value */
+		"cgetaddr	$t1, %2\n\t"
+		"and		$t0, $t1, $t0\n\t"
+		"csetaddr	%0, %0, $t0\n\t"
+		"cscc		$t0, %0, %1\n\t" /* attempt to store */
+		"beqz		$t0, 1b\n\t"	/* spin if failed */
+		"nop\n\t"			/* delay slot */
+		: "=&r" (temp), "+C" (p)
+		: "r" (v)
+		: "t0", "t1", "memory");
+}
+
+static __inline void
+atomic_add_ptr(volatile uintptr_t *p, uintptr_t v)
+{
+	uintptr_t temp;
+
+	__asm __volatile (
+		"1:\n\t"
+		"cllc		%0, %1\n\t"	/* load old value */
+		"cgetaddr	$t0, %2\n\t"	/* calculate new value */
+		"cincaddr	%0, %0, $t0\n\t"
+		"cscc		$t0, %0, %1\n\t" /* attempt to store */
+		"beqz		$t0, 1b\n\t"	/* spin if failed */
+		"nop\n\t"			/* delay slot */
+		: "=&r" (temp), "+C" (p)
+		: "r" (v)
+		: "t0", "memory");
+}
+
+static __inline void
+atomic_subtract_ptr(volatile uintptr_t *p, uintptr_t v)
+{
+	uintptr_t temp;
+
+	__asm __volatile (
+		"1:\n\t"
+		"cllc		%0, %1\n\t"	/* load old value */
+		"cgetaddr	$t0, %2\n\t"	/* calculate new value */
+		"cgetaddr	$t1, %0\n\t"
+		"dsubu		$t0, $t1, $t0\n\t"
+		"csetaddr	%0, %0, $t0\n\t"
+		"cscc		$t0, %0, %1\n\t" /* attempt to store */
+		"beqz		$t0, 1b\n\t"	/* spin if failed */
+		"nop\n\t"			/* delay slot */
+		: "=&r" (temp), "+C" (p)
+		: "r" (v)
+		: "t0", "memory");
+}
+
+#define ATOMIC_ACQ_REL_PTR(NAME)					\
+static __inline  void							\
+atomic_##NAME##_acq_ptr(volatile uintptr_t *p, uintptr_t v)		\
+{									\
+	atomic_##NAME##_ptr(p, v);					\
+	mips_sync(); 							\
+}									\
+									\
+static __inline  void							\
+atomic_##NAME##_rel_ptr(volatile uintptr_t *p, uintptr_t v)		\
+{									\
+	mips_sync();							\
+	atomic_##NAME##_ptr(p, v);					\
+}
+
+ATOMIC_ACQ_REL_PTR(set)
+ATOMIC_ACQ_REL_PTR(clear)
+ATOMIC_ACQ_REL_PTR(add)
+ATOMIC_ACQ_REL_PTR(subtract)
+
+#undef ATOMIC_ACQ_REL_PTR
+
+static __inline int
+atomic_cmpset_ptr(volatile uintptr_t *p, uintptr_t cmpval, uintptr_t newval)
```

Comment:
Can we use 
```
ATOMIC_CMPSET(ptr);
ATOMIC_FCMPSET(ptr);
```
here in the same was as the RISC-V file?
It seems to generate mostly the same code except for some unnecessary shifting: https://cheri-compiler-explorer.cl.cam.ac.uk/z/Y1ofTa

Reply:
The shifting should be fixed by https://reviews.llvm.org/rG34da12dd1fb327bb56f5003a727a73c78929a98d if we care enough to cherry-pick it.